### PR TITLE
verify supported availability zones before instance type changes

### PR DESCRIPTION
**Pull request by aidenkeating**
## Overview

Jira: https://issues.redhat.com/browse/INTLY-9741

currently, the aws provider does not perform a check to confirm
that changes to the instance types of redis instances are actually
supported in the availability zones that the elasticache instances
are using.

this change ensures:

- for updates, that any update to the instance type of elasticache
  instances is actually supported in the availability zones of the
  instance. if not, a modification will not be performed.
- for creates, availability zones are not used if they do not
  support the default instance types set in the cloud resource
  operator.

### Standalone Installation

- In a cluster provisioned in `us-east-1`
- From this branch, simulate a region that doesn't have two AZs that support the instance type e.g.
```
	var tempAzs []*ec2.AvailabilityZone
	for _, az := range azs.AvailabilityZones {
		if aws.StringValue(az.ZoneName) == "us-east-1e" || aws.StringValue(az.ZoneName) == "us-east-1a" {
			tempAzs = append(tempAzs, az)
		}
	}

	// filter the availability zones to only include ones that support the default instance types.
	var supportedAzs []*ec2.AvailabilityZone
	for _, az := range tempAzs {
		for _, instanceTypeOffering := range describeInstanceTypeOfferingsOutput.InstanceTypeOfferings {
			if aws.StringValue(instanceTypeOffering.Location) == aws.StringValue(az.ZoneName) {
				supportedAzs = append(supportedAzs, az)
				continue
			}
		}
	}
```
- From this branch, create a Redis resource (`make cluster/seed/managed/redis`)
- Run the operator (`make run`), ensure it fails with `error creating vpc subnets: expected 2 availability zones`
- Stash/delete the changes
- Run the operator (`make run`) with just the changes in this branch
- Ensure a `cache.t3.micro` Elasticache instance comes up successfully

### Standalone Upgrade

- In a cluster provisioned in `us-east-1`
- From master, in `provider_redis.go` update `defaultCacheNodeType` to `cache.t2.micro`
- From master, update `cluster_network_provider.go` to force `us-east-1e` to be used e.g.
```
	for _, az := range azs.AvailabilityZones {
		if aws.StringValue(az.ZoneName) == "us-east-1e" || aws.StringValue(az.ZoneName) == "us-east-1a" {
			validAzs = append(validAzs, az)
		}
	}
```
- Create a Redis resource (`make cluster/seed/managed/redis`), run the operator from master (`make run`), and wait for it to provision in `us-east-1`
- Ensure at least one of the Elasticache nodes exists in `us-east-1e`
- From master,  in `provider_redis.go` update `defaultCacheNodeType` to `cache.t3.micro`
- Ensure an error `failed to modify elasticache cluster: InsufficientCacheClusterCapacity: cache.t3.micro` is shown and the CR goes into a failure state
- From this branch, run the operator ensure `cache node type cache.t3.micro is not supported, skipping cache node type modification` is logged and the CR is in a valid state

### Integreatly Verification

- After the Standalone verifications have been performed and a release have been cut
- Ensure manifests are added to Integreatly to allow an upgrade path from 2.4.0 and 2.5.0
  - The `0.16.1` manifest should be included, with no `replaces` statement
  - The `0.17.0` manifest should be included, with `replaces` the `0.16.1` manifest
  - The new release manifest should be included, with `replaces` the `0.17.0` manifest
- This verification will need to be performed again for a cherry-pick to the 2.6.0 release branch, for 2.6.1

#### Integreatly Installation

- From master, ensure Integreatly comes up successfully and the version of CRO is the latest manifest version

#### Integreatly Upgrade (2.5.0)

- On an OpenShift cluster provisioned in `eu-east-1`, install Integreatly 2.5.0
- Ensure an upgrade to master successfully finishes with CRO at the latest manifest version
- Ensure all cloud resources are in a successful state

#### Integreatly Upgrade (2.4.0)

- On an OpenShift < 4.4.6 cluster provisioned in `eu-east-1`, install Integreatly 2.4.0
- Ensure an upgrade to master successfully finishes with CRO at the latest manifest version
- Ensure all cloud resources are in a successful state

### CODE REVIEW

File: pkg/providers/aws/credentials.go [hunk 1]
```diff
@@ -62,6 +62,7 @@ var (
 				"ec2:DescribeRouteTables",
 				"ec2:CreateRoute",
 				"ec2:DeleteRoute",
+				"ec2:DescribeInstanceTypeOfferings",
```
Comment (aidenkeating) [inline pkg/providers/aws/credentials.go:65]: The permissions documentation needs to be updated with this new credential and @ciaran-byrne should be made aware of this update, as it'll be required for the next release.

**Comment by aidenkeating** [comment]
Still need to run a full e2e verification against a cluster. I will update verification steps once complete.

**Comment by ciaranRoche** [comment]
# Standalone Installation 
### Simulated a region that doesn't have two AZs
image
### Ran from branch
image
image

**Comment by ciaranRoche** [comment]
## Standalone Upgrade
### Installed from master t2
image
### Failed to modify from master to t3
image
### Skipped modify from this branch
image

**Pull request merged**